### PR TITLE
DAOS-5759 obj: refine DAOS_GET_RECX handling for EC

### PR DESCRIPTION
**Pull request by liuxuezhao**
For EC obj need to translate from VOX extend to DAOS extend.

Signed-off-by: Xuezhao Liu <xuezhao.liu@intel.com>

### CODE REVIEW

File: src/object/cli_shard.c [hunk 1]
```diff
@@ -1362,10 +1362,50 @@ struct obj_query_key_cb_args {
...
+		return;
+	}
+
+	rx_end = reply_recx->rx_idx + reply_recx->rx_nr;
+	rx_idx = max(reply_recx->rx_idx, rounddown(rx_end - 1, cell_rec_nr));
+	result_recx->rx_nr = rx_end - rx_idx;
+	rx_idx = obj_ec_idx_vos2daos(rx_idx, stripe_rec_nr, cell_rec_nr,
+				     tgt_idx);
+	result_recx->rx_idx = rx_idx;
+}
```
Comment (wangdi1) [inline src/object/cli_shard.c:1407]: probably this should be moved to cli_ec.c. merge to obj_recx_ec2_daos()?

**Review by wangdi1 — DISMISSED** [review]

**Review by ravalsam — DISMISSED** [review]
Tried this PR and looks good to me. I do not see any File size WARNING message any more with IOR.

File: src/object/cli_shard.c (hunk 1 above)
Comment (liuxuezhao) [reply src/object/cli_shard.c:1407]: this func() only generate one result recx that is different with obj_recx_ec2_daos(), we can consider to merge or reuse some code for them in following patch.

**Comment by liuxuezhao** [comment]
CI failure due to https://jira.hpdd.intel.com/browse/DAOS-5923

**Comment by liuxuezhao** [comment]
ping ...

File: src/vos/vos_query.c [hunk 2]
```diff
@@ -200,11 +208,26 @@ query_recx(struct open_query *query, daos_recx_t *recx)
 	recx->rx_idx = entry.en_sel_ext.ex_lo;
 	recx->rx_nr = entry.en_sel_ext.ex_hi - entry.en_sel_ext.ex_lo + 1;
```
Comment (jolivier23) [inline src/vos/vos_query.c:209]: Is there an advantage to not masking off the parity indicator at this layer ?

**Review by jolivier23 — APPROVED** [review]

File: src/include/daos/object.h [hunk 3]
```diff
@@ -27,6 +27,13 @@
 #include <daos/tse.h>
 #include <daos_obj.h>
 
+/* EC parity is stored in a private address range that is selected by setting
+ * the most-significant bit of the offset (an unsigned long). This effectively
+ * limits the addressing of user extents to the lower 63 bits of the offset
+ * range.
+ */
+#define DAOS_EC_PARITY_BIT	(1ULL << 63)
```
Comment (wangdi1) [inline src/include/daos/object.h:35]: why define a new variable, instead of using PARITY_INDICATOR?

**Review by wangdi1 — APPROVED** [review]

File: src/include/daos/object.h (hunk 3 above)
Comment (liuxuezhao) [reply src/include/daos/object.h:35]: the naming of "PARITY_INDICATOR" without good prefix seems not good to put in  daos/object.h, actually i think can replay previous usage of PARITY_INDICATOR to DAOS_EC_PARITY_BIT.

File: src/vos/vos_query.c (hunk 2 above)
Comment (liuxuezhao) [reply src/vos/vos_query.c:209]: For EC, it needs to iterate both data space and parity space to know which one is the user wanted recx (MAX or MIN). To hide the details of how to compare/check which one is MAX/MIN (that need some EC detailed logic), in this patch at server-side it iterates both spaces and returns both recxs to client, and client will check which one to return to user.

**Pull request merged**